File: frontend/src/lib/account-types.ts
import type { ElementType } from 'react'
import {
  Building2,
  CircleDashed,
  CreditCard,
  HandCoins,
  PiggyBank,
  TrendingUp,
  Wallet,
} from 'lucide-react'

/**
 * The account-type registry — the frontend half of `app/core/account_types.py`.
 *
 * Same shape and the same reason: the type used to be compared against the
 * literal `'credit_card'` at every site that cared (the create dialog, the icon
 * tile, the import mapper, the balance coloring, the account detail page), so
 * adding a type meant finding all of them. Here a type is one row, and a site
 * asks about the trait it actually depends on.
 *
 * `keys` MUST stay in sync with ACCOUNT_TYPE_KEYS on the backend: the API
 * validates writes against its own list, so a type offered here but missing
 * there fails with a 422 at save time.
 *
 * Traits mirror the backend's, minus the ones only the server acts on
 * (`countsPendingInBalance` has no frontend consumer, so it isn't modeled):
 *
 *   isLiability      Money owed rather than money held. Drives balance coloring
 *                    and the assets-vs-liabilities split in reports.
 *   hasBillingCycle  Bills in statement cycles. Gates the credit-limit and
 *                    close/due-day fields in the dialog, the "due in N days"
 *                    hint, the bills feed, and the cycle-shaped default date
 *                    range on the account detail page.
 */
export interface AccountTypeSpec {
  key: string
  /** i18n key for the display name. */
  labelKey: string
  icon: ElementType
  /** Tailwind text color for the type tile. */
  color: string
  /** Tailwind background for the type tile. */
  bg: string
  isLiability: boolean
  hasBillingCycle: boolean
  /**
   * i18n key for the balance field's label in the create/edit dialog. Debt
   * accounts ask for the amount owed rather than a signed balance, because the
   * API stores liabilities positive-for-debt.
   */
  balanceLabelKey: string
  /** i18n key for the hint under that field, when the type needs one. */
  balanceHintKey?: string
}

// Declaration order is the order the create/edit dialog offers these in:
// grouped by what the account is (cash, then debt, then holdings), not
// alphabetically.
export const ACCOUNT_TYPES: readonly AccountTypeSpec[] = [
  {
    key: 'checking',
    labelKey: 'accounts.typeChecking',
    icon: Building2,
    color: 'text-indigo-600',
    bg: 'bg-indigo-100',
    isLiability: false,
    hasBillingCycle: false,
    balanceLabelKey: 'accounts.balance',
  },
  {
    key: 'savings',
    labelKey: 'accounts.typeSavings',
    icon: PiggyBank,
    color: 'text-emerald-600',
    bg: 'bg-emerald-100',
    isLiability: false,
    hasBillingCycle: false,
    balanceLabelKey: 'accounts.balance',
  },
  {
    key: 'credit_card',
    labelKey: 'accounts.typeCreditCard',
    icon: CreditCard,
    color: 'text-violet-600',
    bg: 'bg-violet-100',
    isLiability: true,
    hasBillingCycle: true,
    balanceLabelKey: 'accounts.balanceCreditCard',
    balanceHintKey: 'accounts.balanceCreditCardHint',
  },
  {
    key: 'loan',
    labelKey: 'accounts.typeLoan',
    icon: HandCoins,
    color: 'text-orange-600',
    bg: 'bg-orange-100',
    isLiability: true,
    hasBillingCycle: false,
    balanceLabelKey: 'accounts.balanceCreditCard',
    balanceHintKey: 'accounts.balanceLoanHint',
  },
  {
    key: 'investment',
    labelKey: 'accounts.typeInvestment',
    icon: TrendingUp,
    color: 'text-amber-600',
    bg: 'bg-amber-100',
    isLiability: false,
    hasBillingCycle: false,
    balanceLabelKey: 'accounts.balance',
  },
  {
    key: 'wallet',
    labelKey: 'accounts.typeWallet',
    icon: Wallet,
    color: 'text-rose-600',
    bg: 'bg-rose-100',
    isLiability: false,
    hasBillingCycle: false,
    balanceLabelKey: 'accounts.balance',
  },
  {
    key: 'other',
    labelKey: 'accounts.typeOther',
    icon: CircleDashed,
    color: 'text-slate-600',
    bg: 'bg-slate-100',
    isLiability: false,
    hasBillingCycle: false,
    balanceLabelKey: 'accounts.balance',
  },
] as const

export const DEFAULT_ACCOUNT_TYPE = 'checking'

const BY_KEY = new Map(ACCOUNT_TYPES.map((s) => [s.key, s]))

/**
 * Resolve a spec, falling back to checking for a type this build doesn't know.
 *
 * The column is free-form server-side and older databases can hold values that
 * predate the registry, so an unknown type has to render as *something* rather
 * than crash the accounts list.
 */
export function getAccountTypeSpec(type: string | null | undefined): AccountTypeSpec {
  return (type ? BY_KEY.get(type) : undefined) ?? BY_KEY.get(DEFAULT_ACCOUNT_TYPE)!
}

export function isLiabilityType(type: string | null | undefined): boolean {
  return getAccountTypeSpec(type).isLiability
}

export function hasBillingCycle(type: string | null | undefined): boolean {
  return getAccountTypeSpec(type).hasBillingCycle
}

/** i18n key → display name, for any site that only needs the label. */
export const ACCOUNT_TYPE_LABEL_KEYS: Record<string, string> = Object.fromEntries(
  ACCOUNT_TYPES.map((s) => [s.key, s.labelKey]),
)

File: backend/app/core/account_types.py
"""The account-type registry: one row per type, one place to add the next.

`Account.type` is a free-form string column, and for most of the product it is
only a label — an icon, a color, a translated name. A few types carry actual
*behavior*, and each behavior used to be spelled as `account.type ==
"credit_card"` at every site that cared: nine separate sites for the sign
convention alone. Adding a second debt-shaped type meant finding all nine, and
missing one produced a quietly wrong number rather than an exception.

So the behaviors are named here as traits, and the call sites ask about the
trait instead of the type. Adding a type is a row in `ACCOUNT_TYPES`; adding a
behavior is a field on `AccountTypeSpec` plus the sites that honor it.

The traits are deliberately independent. `credit_card` happens to set all three
today, but they answer different questions, and the first type to split them
apart is `loan`:

  is_liability
      Does this account represent money owed? Drives the sign convention —
      providers report debt as a positive number and every display site negates
      it — and the assets-vs-liabilities split in net worth.

  counts_pending_in_balance
      Is an authorized-but-uncleared row already real? True for a card, where a
      pending purchase is already owed and the bill will include it. False for a
      loan, whose balance moves only when a payment actually settles.

  has_billing_cycle
      Does this account bill in statement cycles? Drives `effective_date`
      bucketing, bill-vs-P&L totals, and the available-credit / cycle-date
      fields. A loan has a due date but no cycle: nothing accrues to a
      statement, so none of that machinery applies to it.

`metadata_fields` names the columns a type owns. Changing an account's type
clears the columns the new type does not own, so an account is never left half
credit-card. A future type that brings its own columns (loan terms — principal,
rate, maturity — being the obvious next one) lists them here and inherits that
behavior without touching the type-change path.
"""
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class AccountTypeSpec:
    key: str
    is_liability: bool = False
    counts_pending_in_balance: bool = False
    has_billing_cycle: bool = False
    metadata_fields: frozenset[str] = frozenset()


# Columns that only make sense on an account with a statement cycle. Kept as a
# name so the registry row reads as a statement about the type rather than a
# six-line literal.
_CARD_FIELDS = frozenset({
    "credit_limit",
    "statement_close_day",
    "payment_due_day",
    "minimum_payment",
    "card_brand",
    "card_level",
})

# Declaration order is the order the UI offers these in, so it is grouped by
# what the account *is* (cash, then debt, then holdings) rather than
# alphabetically.
ACCOUNT_TYPES: tuple[AccountTypeSpec, ...] = (
    AccountTypeSpec("checking"),
    AccountTypeSpec("savings"),
    AccountTypeSpec(
        "credit_card",
        is_liability=True,
        counts_pending_in_balance=True,
        has_billing_cycle=True,
        metadata_fields=_CARD_FIELDS,
    ),
    AccountTypeSpec("loan", is_liability=True),
    AccountTypeSpec("investment"),
    AccountTypeSpec("wallet"),
    # The catch-all. Present so a type this build doesn't recognize has a real
    # home rather than only the silent `_FALLBACK`, and because the agent-facing
    # MCP contract has advertised it as a filterable value since before the
    # registry existed (mcp_server/tools/transactions.py).
    AccountTypeSpec("other"),
)

_SPECS: dict[str, AccountTypeSpec] = {s.key: s for s in ACCOUNT_TYPES}

# Unknown types resolve to the plainest possible reading: an asset-shaped
# account with no special handling. `type` is a free-form column that predates
# this registry and is written by provider mappers, so a database can hold a
# value this build has never heard of. A KeyError here would turn one stale row
# into a 500 on the whole accounts list.
_FALLBACK = AccountTypeSpec("__unknown__")

ACCOUNT_TYPE_KEYS: tuple[str, ...] = tuple(s.key for s in ACCOUNT_TYPES)
DEFAULT_ACCOUNT_TYPE = "checking"

# Key sets for SQL `IN` clauses, where the filter runs inside the database and
# cannot call a Python predicate per row. Sorted tuples rather than sets so the
# rendered SQL is byte-stable across processes — set iteration order is not, and
# an unstable literal list defeats statement caching and makes query plans
# harder to compare between runs.
LIABILITY_TYPES: tuple[str, ...] = tuple(
    sorted(s.key for s in ACCOUNT_TYPES if s.is_liability)
)
PENDING_IN_BALANCE_TYPES: tuple[str, ...] = tuple(
    sorted(s.key for s in ACCOUNT_TYPES if s.counts_pending_in_balance)
)
BILLING_CYCLE_TYPES: tuple[str, ...] = tuple(
    sorted(s.key for s in ACCOUNT_TYPES if s.has_billing_cycle)
)

# Every column owned by some type — the set `clear_unowned_metadata` clears from.
ALL_METADATA_FIELDS: frozenset[str] = frozenset().union(
    *(s.metadata_fields for s in ACCOUNT_TYPES)
)


def spec(account_or_type: Any) -> AccountTypeSpec:
    """Resolve a spec from an account-like object or a bare type string.

    Accepts either so call sites can pass whatever they already hold — an
    `Account`, an `AccountCreate`, a provider's `AccountData`, or the string
    itself — without an attribute dance at each one.
    """
    key = getattr(account_or_type, "type", account_or_type)
    if not isinstance(key, str):
        return _FALLBACK
    return _SPECS.get(key, _FALLBACK)


def is_liability(account_or_type: Any) -> bool:
    """True when the account represents money owed rather than money held."""
    return spec(account_or_type).is_liability


def counts_pending_in_balance(account_or_type: Any) -> bool:
    """True when authorized-but-uncleared rows already belong in the balance."""
    return spec(account_or_type).counts_pending_in_balance


def has_billing_cycle(account_or_type: Any) -> bool:
    """True when the account bills in statement cycles (bucketing, bills, limit)."""
    return spec(account_or_type).has_billing_cycle


def balance_sign(account_or_type: Any) -> int:
    """The multiplier taking a stored balance to its displayed value.

    Liabilities are stored positive-for-debt (the convention every provider
    reports and `_simplefin_to_internal_balance` normalizes SimpleFIN into) and
    displayed negative. Everything else stores and displays the same number.
    """
    return -1 if is_liability(account_or_type) else 1


def metadata_fields(account_or_type: Any) -> frozenset[str]:
    """The type-owned columns this type is allowed to carry."""
    return spec(account_or_type).metadata_fields


def clear_unowned_metadata(account: Any) -> None:
    """Null every type-owned column the account's current type does not own.

    Safe to call unconditionally: it only clears fields outside the type's own
    set, so an account keeps exactly the metadata its type declares.
    """
    for name in ALL_METADATA_FIELDS - metadata_fields(account):
        setattr(account, name, None)
